src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -552,41 +556,141 @@ def calculate_dense(self):
     def expectation(self, state, normalize=False):
         return Hamiltonian.expectation(self, state, normalize)
 
-    def expectation_from_samples(self, freq, qubit_map=None):
-        terms = self.terms
-        for term in terms:
+    def _exp_from_circuit(
+        self, circuit: "Circuit", qubit_map: dict, nshots: int = None
+    ):
+        """
+        Calculate the expectation value from a circuit.
+        This allows for non diagonal observables. Each term of the observable is
+        treated separately, by measuring in the correct basis and re-executing the
+        circuit.
+
+        Args:
+            circuit (Circuit): input frequencies.
+            qubit_map (dict): qubit map.
+
+        Returns:
+            (float): the calculated expectation value.
+        """
+        from qibo import Circuit, gates
+
+        if nshots is None:  # pragma: no cover
+            nshots = 1000
+        rotated_circuits = []
+        coefficients = []
+        Z_observables = []
+        q_maps = []
+        for term in self.terms:
+            # store coefficient
+            coefficients.append(term.coefficient)
+            # build diagonal observable
+            Z_observables.append(
+                SymbolicHamiltonian(
+                    prod([Z(q) for q in term.target_qubits]),
+                    nqubits=circuit.nqubits,
+                    backend=self.backend,
+                )
+            )
+            # prepare the measurement basis and append it to the circuit
+            measurements = [
+                gates.M(factor.target_qubit, basis=factor.gate.__class__)
+                for factor in set(term.factors)
+            ]
+            circ_copy = circuit.copy(True)
+            [circ_copy.add(m) for m in measurements]
+            rotated_circuits.append(circ_copy)
+            # map the original qubits into 0, 1, 2, ...
+            q_maps.append(
+                dict(
+                    zip(
+                        [qubit_map[q] for q in term.target_qubits],
+                        range(len(term.target_qubits)),
+                    )
+                )
+            )
+        frequencies = [
+            result.frequencies()
+            for result in self.backend.execute_circuits(rotated_circuits, nshots=nshots)
```

Comment:
I understand this is useful only for hardware execution, right? Since in simulation you could just simulate the circuit and use `expectation` directly, which should also be more efficient (and exact).

Reply:
Yes I had exactly in mind hardware execution for this. But even for simulation, you wouldn't want the exact expectation from the final state (is this what you are suggesting? not sure) as this is supposed to be sample dependent.

Reply:
> you wouldn't want the exact expectation from the final state (is this what you are suggesting? not sure) as this is supposed to be sample dependent.

Yes, above I meant to get the exact expectation from the final state, however you can still get a sample dependent estimate by first simulating the circuit with shots, getting the samples and using `expectation_from_samples`. I would expect this to be faster than using `expectation_from_circuit` for simulation, since in the first case the "state preparation" circuit is only simulated once, while `expectation_from_circuit` will repeat the simulation of the same circuit for every basis.

On hardware, or if you are simulating with noise, it is probably a different story.

Reply:
Ah yeah, I see, I've been thinking about doing the state preparation first and then just measuring for simulation. To do that though, we should probably move (at least some part of) the `expectation_from_circuit` inside the backend, which I am not against by the way, but for the moment I decided to keep it simple. In any case, I was assuming that since `execute_circuits` launches several parallel jobs, if I am not mistaken, the overhead of running several copies of the same simulation would be reduced, assuming you have many cores available.

Reply:
I also think we don't need to do anything about this, since it is already available by executing the circuit manually and then calling `expectation_from_samples`. This is a bit more manual for the user, but it is still fairly simple and it is not the best idea to provide multiple ways (almost equally simple) to do the same thing as it would increase maintenance from our side.